### PR TITLE
8315097: Rename createJavaProcessBuilder

### PR DESCRIPTION
**Pull request by lkorinth**
This pull request renames `createJavaProcessBuilder` to `createLimitedTestJavaProcessBuilder` and renames `createTestJvm` to `createTestJavaProcessBuilder`. Both are implemented through a private `createJavaProcessBuilder`. It also updates the java doc.

This is so that it should be harder to by mistake use `createLimitedTestJavaProcessBuilder` that is problematic because it will not forward JVM flags to the tested JVM.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (3 reviews required, with at least 3 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8315097](https://bugs.openjdk.org/browse/JDK-8315097): Rename createJavaProcessBuilder (**Sub-task** - P4)

### Reviewers
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**) ⚠️ Review applies to [2f57a32d](https://git.openjdk.org/jdk/pull/15452/files/2f57a32df8d17da51a04177563327ca2a75e8061)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [4b2d1711](https://git.openjdk.org/jdk/pull/15452/files/4b2d171133c40c5c48114602bfd0d4da75531317)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [2f57a32d](https://git.openjdk.org/jdk/pull/15452/files/2f57a32df8d17da51a04177563327ca2a75e8061)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15452/head:pull/15452` \
`$ git checkout pull/15452`

Update a local copy of the PR: \
`$ git checkout pull/15452` \
`$ git pull https://git.openjdk.org/jdk.git pull/15452/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15452`

View PR using the GUI difftool: \
`$ git pr show -t 15452`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15452.diff">https://git.openjdk.org/jdk/pull/15452.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15452#issuecomment-1695950412)

### CODE REVIEW

**Review by lmesnik — APPROVED** [review]
The. changes looks good. Please update copyrights for changed filed.
I expect that you completed execution of all tests to ensure that nothing is broken.

**Review by dholmes-ora — APPROVED** [review]
I think the admonition about using this method is a bit strong in that it is natural to use this plain process builder method when a test is going to set its own specific flags for the exec'd process. But I'm okay with renaming to avoid copy'n'paste errors that accidentally use the wrong version.

Thanks

**Review by RogerRiggs — CHANGES_REQUESTED** [review]
I don't think this is the best change across so many files.
It gives a very ugly name to a common test function and affects a very large number of tests.

**Comment by RogerRiggs** [comment]
> @RogerRiggs If it is only the name you want changed, maybe you can offer a better name?
@lkorinth 

Sorry for the too short comment; I wanted to make sure it wasn't integrated before I could look at it more closely.
Neither the bug report or the PR describe the problem and its ramifications, only the solution.
Can you elaborate on the conditions that lead to this. (and include them in the bug report).

**Comment by plummercj** [comment]
> I kind of like that it is long and clumsy, that makes it harder to use...

...but it's used in 462 files. That implies it is commonly used. Are you suggesting nearly all those uses are incorrect and eventually should be converted to `createTestJvm()`?

**Comment by dholmes-ora** [comment]
@RogerRiggs , @plummercj , please see the additional discussion in the parent bug: https://bugs.openjdk.org/browse/JDK-8314823

**Comment by msheppar** [comment]
I don't think  a name change is required. The method is createJavaProcessBuilder with a "list" of argurments and a builder is returned. As such, there is no inference, in the name, that test args will be propagated. Adding the additional java doc description should be sufficient to dispell any misconceptions. 
The createTestJvm is a misnomer as it returns a ProcessBulder rather than a Process, which is what I would expected from createTestJvm, without looking at its signature.

So you could create a single createJavaProcessBuilder with add an additional parameter boolean addTestOpts e.g.
createJavaProcessBuilder(List<String> command, boolean addTestOpts) { ... }

createProcessBuilderIgnoreJavaTestOpt(cmdArgs)  maps to createJavaProcessBuilder(cmdArgs, false)

createTestJvm(cmdArgs) maps to createJavaProcessBuilder(cmdArgs, true)

But this is a lot more work.

alternatively change createTestJvm to createTestJavaProcessBuilder  or createJavaProcessBuilderAddTestOpts

**Comment by RogerRiggs** [comment]
In the context of the goal is to declaratively identify tests that do or do not benefit from additional test flags renaming the `createjavaProcessBuilder` method does not further that goal.

The method name and javadoc of `createjavaProcessBuilder` do not imply that the test options are consulted or used; it only says it creates a ProcessBuilder, and does not promise or document more than that.  The javadoc should probably describe the use of the the three properties that modify the way that the java is launched.

**Comment by lkorinth** [comment]
I have some ideas that I will work on that might lead to a new proposal.

**Comment by stefank** [comment]
From talking to other HotSpot devs it is quite clear that the different names lead to mistakes when writing (copy-n-pasting) tests, so I'm happy that we try to figure out some way to make it more obvious when we prepend the extra test options and when we don't.

I agree with @msheppar that `createTestJvm` isn't a good name and I wouldn't be opposed to changing that name instead of `createJavaProcessBuilder`. However, if we do rename that function I strongly urge us to also rename the corresponding `executeTestJvm` function.

I also think it is too obscure that functions with *Test* in the name prepend the extra test options, and those without *Test* don't, so I'd like to get rid of that convention. 

I wouldn't be opposed to a change that:
* Keeps the `createJavaProcessBuilder` name
* Renames `createTestJvm` to `createJavaProcessBuilderPrependTestOpts`
* Renames `executeTestJvm` to `executeJavaPrependTestOpts`
* Removes `createTestJava`
* Removes `executeTestJava`

**Comment by dholmes-ora** [comment]
> So you could create a single createJavaProcessBuilder with add an additional parameter boolean addTestOpts e.g.
createJavaProcessBuilder(List command, boolean addTestOpts) { ... }

@msheppar that is actually where we started, and it was then split into two differently named methods to "make it clear" which one included the test opts without having to remember the name of the parameter that the true/false argument was bound to.

**Comment by AlanBateman** [comment]
> I wouldn't be opposed to a change that:
> 
> * Keeps the `createJavaProcessBuilder` name
> * Renames `createTestJvm` to `createJavaProcessBuilderPrependTestOpts`
> * Renames `executeTestJvm` to `executeJavaPrependTestOpts`
> * Removes `createTestJava`
> * Removes `executeTestJava`

The good thing with this is that it very explicit at the use-sites where the test opts are prepended. Also moving from "Jvm" to "Java" is probably right as it executes the "java" launcher, the options are a mix of java launcher options (that the VM doesn't know about) and VM options.

**Comment by msheppar** [comment]
> > So you could create a single createJavaProcessBuilder with add an additional parameter boolean addTestOpts e.g.
> > createJavaProcessBuilder(List command, boolean addTestOpts) { ... }
> 
> @msheppar that is actually where we started, and it was then split into two differently named methods to "make it clear" which one included the test opts without having to remember the name of the parameter that the true/false argument was bound to.

cheers David  thanks for the clarification on background

**Comment by lkorinth** [comment]
I have created an alternative that uses enums to force the user to make a decision: https://github.com/openjdk/jdk/compare/master...lkorinth:jdk:+process_tools . Another alternative is to do the same but instead using an enum (I think it is not as good). A third alternative is to use the current pull request with a better name.

What do you prefer? Do you have a better alternative? Do someone still think the current code is good? I think what we have today is inferior to all these improvements, and I would like to make it harder to develop bad test cases.

**Comment by RogerRiggs** [comment]
> What do you prefer? Do you have a better alternative? Do someone still think the current code is good? I think what we have today is inferior to all these improvements, and I would like to make it harder to develop bad test ca

The current API (name) is fine and fit for purpose; it does not promise or hide extra functionality under a simple name.

There needs to be an explicit intention in the test(s) to support after the fact that arbitrary flags can be added.
@AlanBateman's proposal for naming [above](https://github.com/openjdk/jdk/pull/15452#issuecomment-1700459277) (or similar) would capture more clearly that test options are propagated to the child process.
Every test writer should be aware that additional command line options may be mixed in.

There are many cases in which the ProcessTools APIs are not used to create child processes and do not need to be used in writing tests. They provide some convenience but also add a dependency and another API layer to work through in the case of failures.

As far as I'm aware, there is no general guidance or design pattern outside of hotspot tests to propagate flags or use ProcessTools. Adding that as a requirement will need a different level of communication and change.

**Comment by msheppar** [comment]
> From talking to other HotSpot devs it is quite clear that the different names lead to mistakes when writing (copy-n-pasting) tests, so I'm happy that we try to figure out some way to make it more obvious when we prepend the extra test options and when we don't.
> 
> I agree with @msheppar that `createTestJvm` isn't a good name and I wouldn't be opposed to changing that name instead of `createJavaProcessBuilder`. However, if we do rename that function I strongly urge us to also rename the corresponding `executeTestJvm` function.
> 
> I also think it is too obscure that functions with _Test_ in the name prepend the extra test options, and those without _Test_ don't, so I'd like to get rid of that convention.
> 
> I wouldn't be opposed to a change that:
> 
>     * Keeps the `createJavaProcessBuilder` name
> 
>     * Renames `createTestJvm` to `createJavaProcessBuilderPrependTestOpts`
> 
>     * Renames `executeTestJvm` to `executeJavaPrependTestOpts`
> 
>     * Removes `createTestJava`
> 
>     * Removes `executeTestJava`

I think @stefank made a reasonable suggestion which was endorsed by @AlanBateman 

which would remove the misconception hurdle

**Comment by lkorinth** [comment]
I think you are missing the point. If you take a look at [the parent bug of the sub task](https://bugs.openjdk.org/browse/JDK-8314823) you can see that the problem described is *not* that people are using `createTestJvm` in error. The problem is that they are (or possibly are) using `createJavaProcessBuilder` in error. Thus renaming `createTestJvm` might help a little at most for this specific problem. Renaming `createJavaProcessBuilder` most probably helps *more*. I guess the alternative of forcing the user to make a choice using an enum value will help even more.

**Comment by lkorinth** [comment]
sorry...

**Comment by lkorinth** [comment]
Hi, if you want to see what I have modified manually, you can do my sed commands and compare to this pull request:
```
git switch -c _reproduce 15acf4b8d7cffcd0d74bf1b9c43cde9acaf31ea9

find -name "*.java" | xargs -n 1 sed -i -e "s/createJavaProcessBuilder(/createLimitedTestJavaProcessBuilder(/g"
find -name "*.java" | xargs -n 1 sed -i -e "s/createTestJvm(/createTestJavaProcessBuilder(/g"
find -name "*.java" | xargs -n 1 sed -i -e "s/import static jdk.test.lib.process.ProcessTools.createJavaProcessBuilder/import static jdk.test.lib.process.ProcessTools.createLimitedTestJavaProcessBuilder/g"
find -name "*.java" | xargs -n 1 sed -i -e "s/import static jdk.test.lib.process.ProcessTools.createTestJvm/import static jdk.test.lib.process.ProcessTools.createTestJavaProcessBuilder/g"
git add -u; git commit -m sed

git diff-tree --no-commit-id --name-only -r 15acf4b8d7cffcd0d74bf1b9c43cde9acaf31ea9..HEAD | xargs sed -i -e "s%^\( \* Copyright (c) ....\)[^[:alpha:]]*\(Oracle.*\)%\1, 2023, \2%"
git ls-files -m | xargs sed -i -e "s%\(Copyright (c) 2023\), 2023, \(Oracle.*\)%\1, \2%"
git add -u; git commit -m copyright

git diff HEAD 2f57a32df8d17da51a04177563327ca2a75e8061
```

It will give you an easier way to review.

**Review by stefank — APPROVED** [review]

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 1]
```diff
@@ -512,8 +502,8 @@ private static void printStack(Thread t, StackTraceElement[] stack) {
      * @param command Arguments to pass to the java command.
      * @return The ProcessBuilder instance representing the java command.
      */
-    public static ProcessBuilder createTestJvm(List<String> command) {
-        return createTestJvm(command.toArray(String[]::new));
+    public static ProcessBuilder createTestJavaProcessBuilder(List<String> command) {
+        return createTestJavaProcessBuilder(command.toArray(String[]::new));
```
Comment (RogerRiggs) [inline test/lib/jdk/test/lib/process/ProcessTools.java:506]: The javadoc shoul d describe all of the options being added to the ProcessBuilder.
They were inadequated described previously and still are.
The other options (seem to be from the code), test.noclasspath, java.class.path, and test.thread.factory.
The description of test.thread.factory and addTestThreadFactoryArgs method seems inadequately described.

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 2]
```diff
@@ -527,10 +517,54 @@ public static ProcessBuilder createTestJvm(List<String> command) {
...
      */
-    public static ProcessBuilder createTestJvm(String... command) {
+    public static ProcessBuilder createTestJavaProcessBuilder(String... command) {
         return createJavaProcessBuilder(Utils.prependTestJavaOpts(command));
     }
 
+    /**
+     * Create ProcessBuilder using the java launcher from the jdk to
+     * be tested.
+     *
```
Comment (RogerRiggs) [inline test/lib/jdk/test/lib/process/ProcessTools.java:527]: As above, should described the limited options that are added to the ProcessBuilder, the same as for `reateTestJavaProcessBuilder(...)`

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 3]
```diff
@@ -527,10 +517,54 @@ public static ProcessBuilder createTestJvm(List<String> command) {
...
+     * @return The ProcessBuilder instance representing the java command.
+     */
+    public static ProcessBuilder createLimitedTestJavaProcessBuilder(List<String> command) {
+        return createLimitedTestJavaProcessBuilder(command.toArray(String[]::new));
+    }
+
+   /**
+     * Create ProcessBuilder using the java launcher from the jdk to
+     * be tested.
+     *
```
Comment (RogerRiggs) [inline test/lib/jdk/test/lib/process/ProcessTools.java:549]: As above, should described the limited options that are added to the ProcessBuilder, the same as for reateTestJavaProcessBuilder(...)

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 4]
```diff
@@ -562,7 +596,7 @@ public static OutputAnalyzer executeTestJvm(List<String> cmds) throws Exception
      * @return The output from the process.
      */
     public static OutputAnalyzer executeTestJvm(String... cmds) throws Exception {
-        ProcessBuilder pb = createTestJvm(cmds);
+        ProcessBuilder pb = createTestJavaProcessBuilder(cmds);
```
Comment (RogerRiggs) [inline test/lib/jdk/test/lib/process/ProcessTools.java:599]: This should also describe *all* of the options being set in the ProcessBuilder before executing the process.

**Comment by stefank** [comment]
@RogerRiggs You seem to know what you want w.r.t. the extra java doc comments. Could you help write those? Could we also do that as a separate RFE? I think that would make it easier to get this PR and the javadoc update through the door.

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 5]
```diff
@@ -527,10 +517,54 @@ public static ProcessBuilder createTestJvm(List<String> command) {
...
+     * createTestJavaProcessBuilder() instead of this method because
+     * createTestJavaProcessBuilder() will add JVM options from
+     * "test.vm.opts" and "test.java.opts"</b> and this method will
+     * not do that.
+     *
+     * <p>If you still chose to use
+     * createLimitedTestJavaProcessBuilder() you should probably use
+     * it in combination with <b>@requires vm.flagless</b> JTREG
+     * anotation as to not waste energy and test resources.
+     *
```
Comment (RogerRiggs) [inline test/lib/jdk/test/lib/process/ProcessTools.java:538]: Consider adding this description of what this method does.
```suggestion
     * Unless the "test.noclasspath" property is "true"
     * the classpath property "java.class.path" is appended to the command line and
     * the environment of the ProcessBuilder is modified to remove "CLASSPATH".
     * If the property "test.thread.factory" is provided the command args are
     * updated and appended to invoke ProcessTools main() and provide the 
     * name of the thread factory.
     * The "-Dtest.thread.factory" is appended to the arguments with the thread factory value.
     * The remaining command args are scanned for unsupported options and 
     * are appended to the ProcessBuilder. 

```

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 6]
```diff
@@ -527,10 +517,54 @@ public static ProcessBuilder createTestJvm(List<String> command) {
...
```
Comment (RogerRiggs) [inline test/lib/jdk/test/lib/process/ProcessTools.java:560]:
```suggestion
     * Unless the "test.noclasspath" property is "true"
     * the classpath property "java.class.path" is appended to the command line and
     * the environment of the ProcessBuilder is modified to remove "CLASSPATH".
     * If the property "test.thread.factory" is provided the command args are
     * updated and appended to invoke ProcessTools main() and provide the 
     * name of the thread factory.
     * The "-Dtest.thread.factory" is appended to the arguments with the thread factory value.
     * The remaining command args are scanned for unsupported options and 
     * are appended to the ProcessBuilder. 
```

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 7]
```diff
@@ -527,10 +517,54 @@ public static ProcessBuilder createTestJvm(List<String> command) {
      * @param command Arguments to pass to the java command.
      * @return The ProcessBuilder instance representing the java command.
      */
-    public static ProcessBuilder createTestJvm(String... command) {
+    public static ProcessBuilder createTestJavaProcessBuilder(String... command) {
```
Comment (RogerRiggs) [inline test/lib/jdk/test/lib/process/ProcessTools.java:520]: Include the same description of other properties that are included in creating the ProcessBuilder...
```
     * Unless the "test.noclasspath" property is "true"
     * the classpath property "java.class.path" is appended to the command line and
     * the environment of the ProcessBuilder is modified to remove "CLASSPATH".
     * If the property "test.thread.factory" is provided the command args are
     * updated and appended to invoke ProcessTools main() and provide the 
     * name of the thread factory.
     * The "-Dtest.thread.factory" is appended to the arguments with the thread factory value.
     * The remaining command args are scanned for unsupported options and 
     * are appended to the ProcessBuilder. 
```

File: test/lib/jdk/test/lib/process/ProcessTools.java [hunk 8]
```diff
@@ -512,8 +502,8 @@ private static void printStack(Thread t, StackTraceElement[] stack) {
      * @param command Arguments to pass to the java command.
      * @return The ProcessBuilder instance representing the java command.
      */
-    public static ProcessBuilder createTestJvm(List<String> command) {
-        return createTestJvm(command.toArray(String[]::new));
+    public static ProcessBuilder createTestJavaProcessBuilder(List<String> command) {
```
Comment (RogerRiggs) [inline test/lib/jdk/test/lib/process/ProcessTools.java:505]: Include the same description of other properties that are included in creating the ProcessBuilder...
```     * Unless the "test.noclasspath" property is "true"
     * the classpath property "java.class.path" is appended to the command line and
     * the environment of the ProcessBuilder is modified to remove "CLASSPATH".
     * If the property "test.thread.factory" is provided the command args are
     * updated and appended to invoke ProcessTools main() and provide the 
     * name of the thread factory.
     * The "-Dtest.thread.factory" is appended to the arguments with the thread factory value.
     * The remaining command args are scanned for unsupported options and 
     * are appended to the ProcessBuilder. 
```

**Review by RogerRiggs — COMMENTED** [review]
Suggestions to complete the descriptions of the createXXXJavaProcessBuilder methods.

**Comment by lkorinth** [comment]
Would it be okay if we handle the new method documentation in a separate pull request? 

After applying your changes, I also noted that the existing description `The command line will be like: {test.jdk}/bin/java {test.vm.opts} {test.java.opts} cmds` is not only incorrect (or at least incomplete), but now also clashes with the added description. I then removed the sentence, but after I did that I also found out that similar wording exist in `executeTestJvm` and I fear that if I continue to pull strings, I will create more and more changes that you will have opinions on.

Is it all right if we push what we have now, and that I create a new pull requests with these improvements in documentation that are actually not related to the changes in this pull request?

**Review by RogerRiggs — APPROVED** [review]
ok, to correct the javadoc in a subsequent PR.

**Comment by lkorinth** [comment]
Thanks, see:
https://bugs.openjdk.org/browse/JDK-8318962

**Pull request closed without merging**